### ISSUE TITLE
[方針] 自己対戦RLの実験結果とActor-Criticへの発展

### ISSUE THREAD

**Issue by eletim** [eletim/napoleon-web#14]
## 目的

教師あり学習したplaying policyを初期値として自己対戦RLを行い、`RuleBasedAgent`の模倣を超える打牌AIを作る。

RL v1としてpure REINFORCEの閉ループを実装し、実際に**1000 iteration / 200,000 self-play games規模**まで学習を実施した。

このIssueでは、その結果と問題点を記録し、次のRL方式への方針を管理する。

> このIssue自体では実装しない。具体的な変更は子`[実装]` Issueへ分割する。

---

## RL v1 構成

対象はplaying phaseのみ。

```text
bidding           RuleBased
choosing-adjutant RuleBased
exchanging        RuleBased
playing           learned policy
```

自己対戦では5席すべて同じ最新playing policyを共有する。

```text
policy-vN x 5
  -> 200 games
  -> fresh trajectory
  -> REINFORCE 1 epoch
  -> policy-vN+1
```

同じtrajectoryを次世代へ使い回さないon-policy型のループ。

### policy

- MLP
- input: playing model input
- output: 53 card logits
- legal masked categorical sampling
- temperature 1.0

### reward

全role共通:

```text
win  = +1
lose = -1
```

reward shapingなし。

### REINFORCE

概念的に:

```text
L = - mean(R * log pi(a|s))
```

最終勝敗を、そのplayerがそのゲーム中に行った全playing decisionsへ付与する。

### iteration標準値

- 200 self-play games / iteration
- 約10,000 playing samples / iteration
- batch size 128
- 1 epoch
- learning rate 1e-5
- 約79 optimizer steps / iteration

---

## 実装済み基盤

- #72 playing self-play trajectory generator
- #74 Python self-play loader
- #76 pure REINFORCE trainer
- #78 / PR #79 iterative RL orchestrator

orchestratorは:

- checkpoint -> ONNX export
- fresh self-play
- REINFORCE
- next checkpoint
- resume
- artifact SHA tracking
- 固定seed periodic evaluation

を自動実行する。

---

# 1000 iteration 実験

## 規模

合計:

```text
1000 iterations
x 200 games
= 200,000 self-play games
```

playing decisionsは約:

```text
10,000 samples / iteration
x 1000
= 約10,000,000 decisions
```

optimizer updatesは約79,000回。

最初の100 iterationと、その後の900 iterationはrun directoryを分けて実施したが、後半runは前半v100 checkpointを初期値としているため、global generationとしてv0〜v1000まで連続している。

---

## 評価条件

10 generationごとに固定seed評価。

1 checkpointあたり:

```text
100 seeds x 5-seat rotation
= 500 games
```

対戦条件:

```text
evaluated seat:
  playing = learned ONNX policy
  other phases = RuleBased

other 4 seats:
  all phases = RuleBased
```

したがって評価勝率は、**playingだけlearned policyへ置き換えた1席 vs RuleBased x4** の勝率。

全generationで同じevaluation seed集合を使う。

---

# 主結果

代表点:

| global generation | win rate vs RuleBased x4 | delta vs supervised v0 |
|---:|---:|---:|
| v0 | 51.8% | 0.0pt |
| v100 | 54.0% | +2.2pt |
| v270 | 56.2% | +4.4pt |
| v500 | 56.0% | +4.2pt |
| v750 | 56.6% | +4.8pt |
| **v840** | **57.0%** | **+5.2pt** |
| v1000 | 54.8% | +3.0pt |

### 解釈

pure REINFORCEでも学習信号は明確に存在した。

- 最初の100世代で51.8% -> 54.0%
- その後もゆっくり改善
- v600〜v850付近は56%前後の高い領域
- v840で評価上57.0%を記録
- その後は劣化し、v1000では54.8%

したがって:

> 「100世代で完全に頭打ち」ではない。

一方で:

> 「最終世代ほど強くなる」わけでもなく、後半にはpolicy degradationが発生した。

現時点のpure REINFORCE best候補はglobal v840。

---

## role別に見えた傾向

確認済み代表値:

| role | supervised v0 | global v100 | global v1000 |
|---|---:|---:|---:|
| Napoleon | 25.0% | 30.0% | 28.0% |
| Adjutant | 38.2% | 37.1% | 39.3% |
| Alliance | 64.3% | 66.6% | 67.8% |

role別sample countは均等ではないため、数字のCIは広い。

ただし、全体勝率が上昇した一方で、Napoleon側が一貫して同程度伸びているわけではない。Alliance側の改善が比較的目立つ。

また全ゲームのcontract success rateは:

```text
v0     34.0%
v840   33.2%
v1000  32.6%
```

であり、全体win rate上昇とNapoleon-side contract success上昇は一致していない。

※ `contractSuccessRate`は「評価AIがNapoleonだった場合の勝率」そのものではなく、評価対象ゲーム全体の契約成功率なのでrole win rateとは区別する。

---

# 1000 iterationで分かった問題

## 1. 自役割がmodel inputへ明示されていない

現状playing observationには:

- Napoleon player
- revealed adjutant
- called adjutant card
- self hand

等があり、MLPは自分の役割を**推論可能**。

しかし、自役割自体は明示していない。

打牌中の役割はゲーム中固定であり、戦略を大きく条件付ける。

必要な4状態:

1. Napoleon
2. Adjutant
3. Alliance
4. Napoleon + Adjutant（独立副官なし / Napoleon-solo）

MLPへ間接推論を要求するより、role条件を明示した方が自然。

→ #88

---

## 2. pure REINFORCEのcredit assignment / varianceが粗い

現状:

```text
win  -> そのplayerの全手 +1
lose -> そのplayerの全手 -1
```

例:

```text
1〜8手目: 良い
9手目:    致命的ミス
10手目:   普通
result:   lose
```

でも全手が同じ-1で更新される。

逆に悪い手でもteam winなら+1になる。

大量self-playでノイズは平均化できるが、方策がある程度強くなると改善信号が弱くなりやすい。

またv840付近をpeakに後半劣化したことから、更新のvariance / stability改善が必要。

次はCriticで状態価値を推定し、期待値との差でActorを更新する。

```text
V(s) = expected terminal reward
A    = R - V(s)

L_actor = -A * log pi(a|s)
L_value = (R - V(s))^2
```

初版はMonte-Carlo terminal returnを教師とするlearned value baselineとし、PPO/GAEまでは入れない。

→ #89

---

## 3. role sample imbalanceの可能性

通常ゲームでは概ね:

```text
Napoleon  1
Adjutant  1
Alliance  3
```

なので、共有policyへ単純sample mean lossを適用するとAlliance samplesのgradient量が多くなりやすい。

対策候補:

```text
L = mean(
  L_napoleon,
  L_adjutant,
  L_alliance,
  L_napoleon_solo
)
```

のようなrole-balanced loss。

ただし今回は**一旦導入しない**。

まず:

1. role明示
2. Actor-Critic / learned value baseline

だけで再実験し、それでもAllianceだけ伸びる傾向が残る場合に独立Issueとして検討する。

---

## reward weightingについて

Napoleon等の少数roleについて:

```text
win rewardを+2や+3にする
```

案も検討した。

しかし、playing中roleは固定であり、roleごとのreward倍率変更は実質的にrole gradient weightingとして働く。

またreward倍率を変えると「ゲームに勝つ」という目的そのものへ人為的なrole weightingを入れることになる。

REINFORCE / Actor-Criticにおいて5player rewardのsumを0へする数学的必要はない。

したがって次実験では:

```text
all roles:
win  +1
lose -1
```

を維持する。

---

# 次フェーズ

実装順:

- [x] #88 `[実装] 打牌AIの自役割4状態を観測へ明示する`
- [x] #89 `[実装] 打牌RLをlearned value baseline付きActor-Criticへ拡張する`

#89は#88完了後。

その後、新runで500〜1000 iteration程度の学習曲線を取得する。

比較対象:

- supervised v0
- pure REINFORCE best候補 global v840
- role-aware Actor-Critic

見る指標:

- overall win rate
- Napoleon win rate
- Adjutant win rate
- Alliance win rate
- Napoleon-solo win rate
- average point cards
- contract success
- paired fixed-seed outcomes
- Critic value / advantage統計

---

# 後続候補

Actor-Critic実験後に必要なら個別Issue化:

- role-balanced loss
- advantage normalization
- entropy bonus
- PPO clipping
- GAE / TD
- opponent pool / historical policy対戦
- reward shaping
- bidding / adjutant / exchange RL

---

## 完了条件

この方針Issueとしては:

- pure REINFORCE 1000 iteration実験結果を記録
- 問題点を整理
- role明示を実装
- learned value baseline付きActor-Criticを実装
- 新方式で固定seed評価付きself-play実験を実施
- pure REINFORCE bestとの比較結果を記録

まで進める。

## 関連

- 親方針: #15
- 方策学習: #10
- 対戦評価: #11
- 推論統合: #12
- full-policy評価: #70
- self-play trajectory: #72
- Python self-play loader: #74
- REINFORCE: #76
- RL orchestrator: #78
- role明示: #88
- Actor-Critic: #89
